Add Nobø Ecohub integration

oyvindwe opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Add integration for climate device Nobø Ecohub.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/17915

This PR is based on the HACS `hanobo` component: https://github.com/echoromeo/hanobo/

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [x] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

oyvindwe committed:
Initial version of Nobø Ecohub.

oyvindwe committed:
Initial version of Nobø Ecohub.

oyvindwe committed:
Options update listener for Nobø Ecohub

oyvindwe committed:
Unit test for nobo_hub config flow

oyvindwe committed:
Cleanup

oyvindwe committed:
Moved comment re backwards compatibility

oyvindwe committed:
Improved tests

oyvindwe committed:
Improved tests

oyvindwe committed:
Improved tests

oyvindwe committed:
Options flow test
Pylint

oyvindwe committed:
Fix backwards compatibility mode

oyvindwe committed:
Don't require Python 3.9

gjohansson-ST commented:
If there is a version in HACS with same domain that will always take priority over the built-in so a migration will not work as it would still use the hacs one.
If you are the maintainer of the one on HACS that has to go at the same time as this one makes it into HA and then get people to uninstall it (as custom component)

oyvindwe commented:
> If there is a version in HACS with same domain that will always take priority over the built-in so a migration will not work as it would still use the hacs one.
> If you are the maintainer of the one on HACS that has to go at the same time as this one makes it into HA and then get people to uninstall it (as custom component)

Thanks for the tip! I'll coordinate this with the HACS maintainer; @echoromeo

oyvindwe committed:
Import form configuration.yaml

oyvindwe committed:
Check if device is already configured.
Correct tests for only discovering serial prefix
Fix importing when only serial suffix is configured

oyvindwe committed:
Use constants

echoromeo commented:
Hi, So I need to remove the integration from both of these repos?

https://github.com/hacs/default/blob/master/integration
https://github.com/home-assistant/wheels-custom-integrations/blob/master/components/nobo_hub.json

oyvindwe commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,407 @@
+                self._nobo.API.OVERRIDE_TARGET_ZONE,
+                self._id,
+            )
+            # TODO: override to program if new operation mode == current week profile status
```
@echoromeo How did you intend to do this?

oyvindwe replied:
I figured out a way to do it, but when looking at the climate UI, I think it makes sense to keep the current behaviour and just remove the todo.

oyvindwe committed:
Pylint and variable name clenaup.

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,88 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Nobø Ecohub from a config entry."""
+
+    # As there currently is no way to import options from yaml
+    # when setting up a config entry, we fallback to adding
+    # the options to the config entry and pull them out here if
+    # they are missing from the options
+    _async_import_options_from_data_if_missing(hass, entry)
+
```
This is a brand new integration there is nothing to import.
```suggestion
```

oyvindwe replied:
The intention was to import config for those using the HACS version, but I understand that's not wanted in the core.

frenck commented on homeassistant/components/nobo_hub/translations/no.json:
```diff
@@ -0,0 +1,68 @@
+{
```
Please do not include additional languages (only the main `strings.json` and `en.json`)

oyvindwe replied:
Deleted. And I've signed up for Lokalise. :)

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    unsubscribe = entry.add_update_listener(options_update_listener)
+    hass.data[DOMAIN][entry.entry_id][UNSUBSCRIBE] = unsubscribe
+
+    _LOGGER.info(
+        "Nobø Ecohub '%s' is up and running on %s:%s", name, ip_address, serial
+    )
+
```
Config entry setup is already logged by the core, I don't think it is really needed to log this again.

```suggestion
```

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,88 @@
+
+    serial = entry.data.get(CONF_SERIAL)
+    ip_address = entry.data.get(CONF_IP_ADDRESS)
+    name = entry.title
```
```suggestion
```

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,88 @@
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+
+    unsubscribe = entry.add_update_listener(options_update_listener)
+    hass.data[DOMAIN][entry.entry_id][UNSUBSCRIBE] = unsubscribe
```
```suggestion
    entry.async_on_unload(entry.add_update_listener(options_update_listener))
```

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    discover = ip_address is None
+    hub = nobo(serial=serial, ip=ip_address, discover=discover, loop=hass.loop)
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = {HUB: hub}
```
Since we don't need to track the unsub anymore, we can just store the hub directly.
```suggestion
    hass.data.setdefault(DOMAIN, {})
    hass.data[DOMAIN][entry.entry_id] = hub
```

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,88 @@
+from homeassistant.const import CONF_COMMAND_OFF, CONF_COMMAND_ON, CONF_IP_ADDRESS
+from homeassistant.core import HomeAssistant, callback
+
+from .const import CONF_SERIAL, DOMAIN, HUB, UNSUBSCRIBE
```
```suggestion
from .const import CONF_SERIAL, DOMAIN
```

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    await hub.stop()
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok:
+        hass.data[DOMAIN][entry.entry_id][UNSUBSCRIBE]()
```
```suggestion
```

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,88 @@
+        hass.data[DOMAIN][entry.entry_id][UNSUBSCRIBE]()
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    _LOGGER.info("Nobø Ecohub '%s' on %s:%s is stopped", name, ip_address, serial)
+
```
```suggestion
```

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    """Unload a config entry."""
+
+    hub = hass.data[DOMAIN][entry.entry_id][HUB]
+    serial = entry.data.get(CONF_SERIAL)
+    ip_address = entry.data.get(CONF_IP_ADDRESS)
+    name = entry.title
```
```suggestion
```

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    serial = entry.data.get(CONF_SERIAL)
+    ip_address = entry.data.get(CONF_IP_ADDRESS)
+    name = entry.title
+    await hub.stop()
+    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
+    if unload_ok:
```
I think we should stop the hub, once all entities are unload (to prevent possible calls from those against a stopped hub)
```suggestion
    unload_ok = await hass.config_entries.async_unload_platforms(entry, PLATFORMS)
    if unload_ok:
        await hub.stop()
```

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,88 @@
+    """Handle options update."""
+    await hass.config_entries.async_reload(config_entry.entry_id)
+
+
+@callback
+def _async_import_options_from_data_if_missing(hass: HomeAssistant, entry: ConfigEntry):
+    options = dict(entry.options)
+    data = {}
+    importable_options = [CONF_COMMAND_OFF, CONF_COMMAND_ON]
+    found = False
+    for key in entry.data:
+        if key in importable_options and key not in options:
+            options[key] = entry.data[key]
+            found = True
+        else:
+            data[key] = entry.data[key]
+
+    if found:
+        hass.config_entries.async_update_entry(entry, data=data, options=options)
```
There is nothing to import.

```suggestion
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+
+_ZONE_NORMAL_WEEK_LIST_SCHEMA = vol.Schema({cv.string: cv.string})
+
+# Legacy configuration to import from HACS version.
+PLATFORM_SCHEMA = cv.PLATFORM_SCHEMA.extend(
+    {
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_IP_ADDRESS, default="discover"): cv.string,
+        vol.Optional(CONF_COMMAND_OFF, default=""): cv.string,
+        vol.Optional(CONF_COMMAND_ON, default={}): _ZONE_NORMAL_WEEK_LIST_SCHEMA,
+    }
+)
```
We don't support custom integrations.
This does not belong in our core codebase.

```suggestion
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+)
+
+
+async def async_setup_platform(
+    hass: HomeAssistant, config, async_add_entities, discovery_info=None
+) -> bool:
+    """Import Nobø Ecohub configuration from YAML."""
+
+    _LOGGER.warning(
+        "Loading Nobø Ecohub via platform setup is depreciated; Please remove it from your configuration"
+    )
+
+    # Only import if we haven't before.
+    config_entry = _async_find_matching_config_entry(hass)
+    if not config_entry:
+        _LOGGER.warning("Importing Nobø Ecohub from configuration.yaml")
+        hass.async_create_task(
+            hass.config_entries.flow.async_init(
+                DOMAIN,
+                context={"source": SOURCE_IMPORT},
+                data=dict(config),
+            )
+        )
+        return True
+
+    return True
+
+
+@callback
+def _async_find_matching_config_entry(hass):
+    for entry in hass.config_entries.async_entries(DOMAIN):
+        if entry.source == SOURCE_IMPORT:
+            return entry
+
+
```
```suggestion
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, config_entry: ConfigEntry, async_add_devices
```
```suggestion
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_devices: AddEntitiesCallback
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+
+    # Setup connection with hub
+    hub = hass.data[DOMAIN][config_entry.entry_id][HUB]
+    await hub.start()
```
Should this have been started in the integration setup instead?

oyvindwe replied:
Yes. That's symmetric with unload. Done.

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+    return None
+
+
+class NoboZone(ClimateEntity):
+    """Representation of a Nobø zone. A Nobø zone consists of a group of physical devices that are controlled as a unity."""
+
```
```suggestion
class NoboZone(ClimateEntity):
    """Representation of a Nobø zone. 
    
    A Nobø zone consists of a group of physical devices that are
    controlled as a unity.
    """
    
    _attr_max_temp = MAX_TEMPERATURE
    _attr_min_temp = MIN_TEMPERATURE
    _attr_precision = PRECISION_TENTHS
    _attr_preset_modes = PRESET_MODES
    _attr_should_poll = False
    _attr_supported_features = SUPPORT_FLAGS
    _attr_temperature_unit = TEMP_CELSIUS
    

```

See: <https://developers.home-assistant.io/docs/core/entity#entity-class-attributes>

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+    @property
+    def unique_id(self) -> str:
+        """Return a unique ID."""
+        return self._unique_id
+
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        return SUPPORT_FLAGS
+
+    @property
+    def should_poll(self):
+        """Return the polling state."""
+        return False
+
+    @property
+    def name(self):
+        """Return the name of the climate device."""
+        return self._name
+
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement."""
+        return TEMP_CELSIUS
+
+    @property
+    def precision(self):
+        """Return the precision of the system."""
+        return PRECISION_TENTHS  # PRECISION_WHOLE
+
+    @property
+    def min_temp(self):
+        """Return the minimum temperature."""
+        return MIN_TEMPERATURE
+
+    @property
+    def max_temp(self):
+        """Return the maximum temperature."""
+        return MAX_TEMPERATURE
+
```
```suggestion
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+        """Return current preset mode."""
+        return self._current_operation
+
+    @property
+    def preset_modes(self):
+        """Return the preset modes, comfort, away etc."""
+        return PRESET_MODES
+
```
```suggestion
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+        """Initialize the climate device."""
+        self._id = zone_id
+        self._nobo = hub
+        self._unique_id = hub.hub_serial + ":" + zone_id
+        self._name = self._nobo.zones[self._id][ATTR_NAME]
+        self._current_mode = HVAC_MODE_AUTO
```
```suggestion
        self._attr_unique_id = hub.hub_serial + ":" + zone_id
        self._attr_name = self._nobo.zones[self._id][ATTR_NAME]
        self._attr_hvac_mode = HVAC_MODE_AUTO
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+        """Set HVAC mode to comfort(HEAT) or back to normal(AUTO)."""
+        if hvac_mode == HVAC_MODE_AUTO:
+            await self.async_set_preset_mode(PRESET_NONE)
+            self._current_mode = hvac_mode
```
```suggestion
            self._attr_hvac_mode = hvac_mode
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+            self._current_mode = hvac_mode
+        elif hvac_mode == HVAC_MODE_HEAT:
+            await self.async_set_preset_mode(PRESET_COMFORT)
+            self._current_mode = hvac_mode
```
```suggestion
            self._attr_hvac_mode = hvac_mode
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+        if self.can_turn_off():
+            if hvac_mode == HVAC_MODE_OFF:
+                await self.async_set_preset_mode(PRESET_NONE)
+                self._current_mode = hvac_mode
```
```suggestion
                self._attr_hvac_mode = hvac_mode
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+        state = self._nobo.get_current_zone_mode(
+            self._id, dt_util.as_local(dt_util.now())
+        )
+        self._current_mode = HVAC_MODE_AUTO
+        self._current_operation = PRESET_NONE
```
```suggestion
        self._attr_hvac_mode = HVAC_MODE_AUTO
        self._attr_preset_mode = PRESET_NONE
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+        self._current_operation = PRESET_NONE
+
+        if state == nobo.API.NAME_OFF:
+            self._current_mode = HVAC_MODE_OFF
```
```suggestion
            self._attr_hvac_mode = HVAC_MODE_OFF
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+        if state == nobo.API.NAME_OFF:
+            self._current_mode = HVAC_MODE_OFF
+        elif state == nobo.API.NAME_AWAY:
+            self._current_operation = PRESET_AWAY
+        elif state == nobo.API.NAME_ECO:
+            self._current_operation = PRESET_ECO
+        elif state == nobo.API.NAME_COMFORT:
+            self._current_operation = PRESET_COMFORT
```
```suggestion
            self._attr_preset_mode = PRESET_AWAY
        elif state == nobo.API.NAME_ECO:
            self._attr_preset_mode = PRESET_ECO
        elif state == nobo.API.NAME_COMFORT:
            self._attr_preset_mode = PRESET_COMFORT
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+                        self._current_mode = HVAC_MODE_HEAT
+
+        self._current_temperature = self._nobo.get_current_zone_temperature(self._id)
+        self._target_temperature_high = int(
+            self._nobo.zones[self._id][ATTR_TEMP_COMFORT_C]
+        )
+        self._target_temperature_low = int(self._nobo.zones[self._id][ATTR_TEMP_ECO_C])
```
```suggestion
        self._attr_target_temperature_high = int(
            self._nobo.zones[self._id][ATTR_TEMP_COMFORT_C]
        )
        self._attr_target_temperature_low = int(self._nobo.zones[self._id][ATTR_TEMP_ECO_C])
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+        """Return the maximum temperature."""
+        return MAX_TEMPERATURE
+
+    @property
+    def target_temperature_high(self):
+        """Return the highbound target temperature we try to reach."""
+        return self._target_temperature_high
+
+    @property
+    def target_temperature_low(self):
+        """Return the lowbound target temperature we try to reach."""
+        return self._target_temperature_low
+
```
```suggestion
```

frenck commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,406 @@
+            return HVAC_MODES
+        return HVAC_MODES_WITHOUT_OFF
+
+    @property
+    def hvac_mode(self):
+        """Return current operation HVAC Mode."""
+        return self._current_mode
+
+    @property
+    def preset_mode(self):
+        """Return current preset mode."""
+        return self._current_operation
+
```
```suggestion
```

frenck commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,237 @@
+    """Handle a config flow for Nobø Ecohub."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_PUSH
```
This isn't used anymore
```suggestion
```

frenck commented on homeassistant/components/nobo_hub/strings.json:
```diff
@@ -0,0 +1,67 @@
+{
+  "title": "Nob\u00f8 Ecohub",
```
I don't think the brand name needs to be translated?

```suggestion
```

frenck reviewed (commented):
Hi there @oyvindwe 👋 

Thanks for opening a PR! I've done an initial pass for review, that will mainly compact the code a bit. It might be needed to rebase this PR onto the latest `dev` branch in order for all suggestions to work.

oyvindwe committed:
Configure override type in options dialog

oyvindwe committed:
Configure override type in options dialog

oyvindwe committed:
Review

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

oyvindwe committed:
Review

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

homeassistant commented:
Hi @GitHub-Action, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#50913).

Once you do that we will be able to review and accept this pull request.

Thanks!

oyvindwe commented:
@frenck Thanks you very much for the review. The code got much more compact when reusing more from `ClimateEntity`. I've fixed all fallouts and updated the tests as well.

frenck commented:
@oyvindwe You are welcome, however, it seems like you have merged branches instead of git rebasing it. This causes unrelated commits to showing up in your PR, please rebase your commits onto the latest dev branch to clean that up.

oyvindwe committed:
Fix tests

oyvindwe commented:
@frenck I just noticed as well. Rebased now!

oyvindwe committed:
Correct disabling off_command and on_commands ("Default" is a hard coded week profile in the hub).

oyvindwe committed:
Improve options dialog

oyvindwe committed:
Formatting

oyvindwe committed:
pyupgrade

Danielhiversen commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,321 @@
+    for zone_id, zone in hub.zones.items():
+        # Replace unicode non-breaking space (used in Nobø Ecohub) with space
+        zone_name = zone[ATTR_NAME].replace("\xa0", " ")
+        if zone_name in command_on_dict:
+            command_on_name = command_on_dict[zone_name]
```
```suggestion
        if zone_name not in command_on_dict:
            continue
        command_on_name = command_on_dict[zone_name]
```
Will save you an indention level

oyvindwe replied:
Fixed - thanks.

Danielhiversen commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,321 @@
+                _LOGGER.warning(
+                    "Can not turn on (or off) zone '%s', because ON week profile was not specified",
+                    zone_name,
+                )
+            else:
+                command_on_id = _get_id_from_name(command_on_name, hub.week_profiles)
```
```suggestion
                )
                continue
            command_on_id = _get_id_from_name(command_on_name, hub.week_profiles)
```

oyvindwe replied:
Fixed - thanks!

Danielhiversen commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,321 @@
+                    continue  # "normal" overrides
+                if (
+                    self._nobo.overrides[override][ATTR_TARGET_TYPE]
+                    == nobo.API.OVERRIDE_TARGET_ZONE
```
```suggestion
                    == nobo.API.OVERRIDE_TARGET_ZONE and self._nobo.overrides[override][ATTR_TARGET_ID] == self._id
```

oyvindwe replied:
Fixed - thanks.

oyvindwe committed:
Incorporated review comments

Danielhiversen commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,322 @@
+        command_off_id = None
+    else:
+        command_off_id = _get_id_from_name(command_off_name, hub.week_profiles)
+        if command_off_id == "" or command_off_id is None:
```
```suggestion
        if command_off_id in ["", None]:
```

oyvindwe replied:
Thanks - fixed.

Danielhiversen commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,322 @@
+
+            # Find ON command (week profile) for the different zones:
+            command_on_dict = config_entry.options.get(CONF_COMMAND_ON)
+            if command_on_dict is None or command_on_dict.keys().__len__ == 0:
```
```suggestion
            if not command_on_dict:
```
I think this should handle both cases

oyvindwe replied:
Thanks - fixed.

Danielhiversen commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,322 @@
+        if zone_name not in command_on_dict:
+            continue
+        command_on_name = command_on_dict[zone_name]
+        if command_on_name is None or command_on_name == "":
```
```suggestion
        if command_on_name in [None,  ""]:
```

oyvindwe replied:
Thanks - fixed.

Danielhiversen commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,322 @@
+            )
+            continue
+        command_on_id = _get_id_from_name(command_on_name, hub.week_profiles)
+        if command_on_id is None or command_on_id == "":
```
```suggestion
        if command_on_id in [None, ""]:
```

oyvindwe replied:
Thanks - fixed.

Danielhiversen commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,322 @@
+
+    async def async_set_preset_mode(self, preset_mode):
+        """Set new zone override."""
+        if self._nobo.zones[self._id][ATTR_OVERRIDE_ALLOWED] == "1":
+            if preset_mode == PRESET_ECO:
```
```suggestion
        if self._nobo.zones[self._id][ATTR_OVERRIDE_ALLOWED] != "1":
            return
        if preset_mode == PRESET_ECO:
```

oyvindwe replied:
Thanks - fixed.

Danielhiversen commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,222 @@
+        errors = {}
+        if user_input is not None:
+            try:
+                serial = user_input.get(CONF_SERIAL)
+                ip_address = user_input.get(CONF_IP_ADDRESS)
+                if serial is None or not len(serial) == 12 or not serial.isdigit():
+                    raise InvalidSerial()
+                if ip_address is not None:
+                    try:
+                        socket.inet_aton(ip_address)
+                    except OSError:
+                        raise InvalidIP() from OSError
+                else:
+                    for (discovered_ip, serial_prefix) in self.discovered_hubs:
+                        if serial.startswith(serial_prefix):
+                            ip_address = discovered_ip
+                            break
+                    if ip_address is None:
+                        raise DeviceNotFound()
+
+                await self.async_set_unique_id(serial)
+                self._abort_if_unique_id_configured()
+
+                # Test connection
+                hub = nobo(
+                    serial=serial, ip=ip_address, discover=False, loop=self.hass.loop
+                )
+                if await hub.async_connect_hub(ip_address, serial):
+                    await hub.close()
+                    await self.async_set_unique_id(serial, raise_on_progress=False)
+                    self._abort_if_unique_id_configured(
+                        reload_on_update=False, updates=user_input
+                    )
+                    return self.async_create_entry(
+                        title=hub.hub_info["name"], data=user_input
+                    )
+                raise CannotConnect()
```
Consider moving this to a validate_input function

oyvindwe replied:
Good idea. I named it `_test_connection`, as it actually tries to connect, not only validates the user input.

oyvindwe committed:
Incorporated review comments.

oyvindwe commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,221 @@
+        else:
+            for (discovered_ip, serial_prefix) in self.discovered_hubs:
+                if serial.startswith(serial_prefix):
+                    ip_address = discovered_ip
```
No, that is not necessary. This are provided by UDP broadcast discovery in `pynobo`. The address that is checked above is user input.

oyvindwe commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,221 @@
+        )
+
+    def _prefill_identifier(self):
+        for (_, serial) in self.discovered_hubs:
+            return serial + "XXX"
+        return ""
```
That doesn't work as `self.discovered_hubs` is a set of tuples, not a list of lists.

oyvindwe committed:
Incorporated second round of review comments.

oyvindwe commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,323 @@
+    # Find OFF command (week profile) to use for all zones:
+    command_off_name = config_entry.options.get(CONF_COMMAND_OFF)
+    command_on_by_id: dict[str, str] = {}  # By default, nothing can be turned on
+    if command_off_name is None or command_off_name == "":
```
This one has to use a list, otherwise `mypy` fails with:
```
homeassistant/components/nobo_hub/climate.py:84: error: Unsupported operand types for in ("Optional[Any]" and "str")  [operator]
```

oyvindwe commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,323 @@
+        if zone_name not in command_on_dict:
+            continue
+        command_on_name = command_on_dict[zone_name]
+        if command_on_name in [None, ""]:
```
Fixed!

oyvindwe commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,323 @@
+        self._id = zone_id
+        self._nobo = hub
+        self._attr_unique_id = hub.hub_serial + ":" + zone_id
+        self._attr_name = self._nobo.zones[self._id][ATTR_NAME]
```
Fixed.

oyvindwe commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,323 @@
+
+    def can_turn_off(self) -> bool:
+        """Return true if heater can turn off and on."""
+        return self._command_on_id is not None and self._command_off_id is not None
```
You are correct - it requires reconfiguration of the integration to be changed, and that will cause a reload of the component. Fixed.

oyvindwe commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,323 @@
+        )
+
+    @callback
+    def update(self):
```
Fixed.

oyvindwe replied:
After changing this, I realised I had introduced a bug in this async version; there is no push event when the preset changes between eco and heat in normal operation (HVAC_MODE_AUTO), so HA need to poll to get this state change.

oyvindwe commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,323 @@
+    def update(self):
+        """Fetch new state data for this zone."""
+        state = self._nobo.get_current_zone_mode(
+            self._id, dt_util.as_local(dt_util.now())
```
You are correct. Fixed.

oyvindwe commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,323 @@
+                    == nobo.API.OVERRIDE_TARGET_ZONE
+                    and self._nobo.overrides[override][ATTR_TARGET_ID] == self._id
+                ):
+                    self._attr_hvac_mode = HVAC_MODE_HEAT
```
Yes. Fixed.

oyvindwe commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,221 @@
+
+        # Test connection
+        hub = nobo(serial=serial, ip=ip_address, discover=False, loop=self.hass.loop)
+        if await hub.async_connect_hub(ip_address, serial):
```
Thanks. Fixed.

oyvindwe commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,221 @@
+        hub = nobo(serial=serial, ip=ip_address, discover=False, loop=self.hass.loop)
+        if await hub.async_connect_hub(ip_address, serial):
+            await hub.close()
+            await self.async_set_unique_id(serial, raise_on_progress=False)
```
Good question about `async_set_unique_id`. I based this on the config flow of some other integration, but I don't remember which one. I agree it seems unnecessary here.

I'll clean up as suggested!

oyvindwe commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,221 @@
+    """Handles options flow for the component."""
+
+    def __init__(self, config_entry: config_entries.ConfigEntry) -> None:
+        """Initializr the options flow."""
```
Thanks! Fixed.

oyvindwe commented on homeassistant/components/nobo_hub/strings.json:
```diff
@@ -0,0 +1,67 @@
+  "config": {
+    "step": {
+      "user": {
+        "title": "Nobø Ecohub",
```
Thanks! Fixed.

oyvindwe committed:
Add polling to discover preset change in HVAC_MODE_AUTO.

oyvindwe committed:
Added tests/components/nobo_hub to CODEOWNERS.

allenporter commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,219 @@
+        if serial is None or not len(serial) == 12 or not serial.isdigit():
+            raise InvalidSerial()
+
+        if ip_address is not None:
```
Is the serial entirely ignored when an ip is specified or is that also validated here somewhere?

oyvindwe replied:
The (complete) serial number is always required to connect. If the IP address isn't provided by the user, the provided serial number is matched with the discovered hubs on the local network.

allenporter commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,219 @@
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_SERIAL, default=default_suggestion): str,
```
Does the _prefill_identifier overwrite the user typed data from the last step when an error is shown or is that preserved?

oyvindwe replied:
Good catch - user input is not preserved on error. I'll look into this.

oyvindwe replied:
Fixed so user input is preserved on errors.

allenporter commented on tests/components/nobo_hub/test_config_flow.py:
```diff
@@ -0,0 +1,260 @@
+        assert result["type"] == "form"
+        assert result["errors"] == {}
+
+        with patch(
```
My impression is that the first patch for discover hubs is only needed on the init step, then the rest of the code does not need to be indent inside of it. That is, i think this can be unindented by one block.

oyvindwe replied:
You are quite correct.

oyvindwe replied:
Fixed in updated tests.

allenporter commented on homeassistant/components/nobo_hub/strings.json:
```diff
@@ -0,0 +1,66 @@
+  "config": {
+    "step": {
+      "user": {
+        "description": "Start by entering the complete serial number (12 digits) of the Nobø Ecohub you want to add. If you have more than one hub, and they have the same serial prefix, you must also specify the IP address. If any devices were automatically found on your network, they are shown below.\n\nIf you cannot see your device or experience any issues, try specifying the device IP address. ...
```
Did you consider making the list of discovered devices selectable during the config flow rather than making the user re-enter the text here? I realize you also want an option for manual entry in case discovery fails, it just seems more user friendly to make auto discovered devices not require any text input.

oyvindwe replied:
Auto discovered devices still need to get the 3 last digits added to their serial number. I guess this is Nobø's security mechanism. Also, in the assumed case you have two hubs on your local network (I don't know why anybody would have that though), and they have the same serial prefix, we're not able to distinguish the two hubs, so then we also need the ip address even though the hub is auto discovered.

oyvindwe replied:
This is fixed in the new config flow.

allenporter commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,326 @@
+        """Close the Nobø Ecohub socket connection when HA stops."""
+        await hub.stop()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _async_close)
```
I think this may not be needed given `async_unload_entry` will stop the hub. Want to confirm that?

(If this is needed, it belongs in __init__.py given it is not specific to the climate platform)

oyvindwe replied:
I tested this, and `async_unload_entry` isn't called when the hub is stopping. However, since the task loop stopped, the integration implicitly disconnects without this code, but IMO that is depending on the library implementation.

Another side effect of above code is that the listener isn't removed if the integration is deactivated, which results in an additional call to `hub.stop()` when stopping HA. I moved this code to `__init__.py` as suggested, and also added code to remove the listener when unloading the component. (I modelled this based on some other components.)

allenporter commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,326 @@
+            await asyncio.sleep(0.5)
+        elif hvac_mode == HVAC_MODE_OFF:
+            _LOGGER.warning(
+                "User tried to turn off zone %s '%s', but this is not configured so this should be impossible",
```
The logic for this could get simpler if at the top of the function there is something like `if hvac_mode not in self.hvac_modes` and it returns and logs a warning. It generalized a bit to also protect against other types of "impossible" cases.

allenporter replied:
You may consider setting `_attr_hvac_modes` in the constructor (just append HVAC_MODE_OFF to the list when needed), and remove `_can_turn_off` and the separate _WITHOUT_OFF list.

oyvindwe replied:
Thanks! Good idea - I'll do this.

oyvindwe replied:
Fixed.

allenporter commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,326 @@
+                    self._command_off_id,
+                )
+            else:
+                await self._nobo.async_update_zone(
```
Why does this only run when `_can_turn_off` is true? It is not obvious reading this why this happens since naively it looks like it supposed to run anytime the hvac mode isn't off.

oyvindwe replied:
`HVAC_MODE_OFF` (if supported) changes the week profile to the "off" profile. On `HVAC_MODE_HEAT` or `HVAC_MODE_AUTO`, we must return to the normal week profile. I think this can be cleaned up to check current HVAC mode first, and only set week profile if necessary.

oyvindwe replied:
Fixed.

allenporter commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,326 @@
+        """Set new target temperature."""
+        low = int(kwargs.get(ATTR_TARGET_TEMP_LOW))
+        high = int(kwargs.get(ATTR_TARGET_TEMP_HIGH))
+        if low > int(self._nobo.zones[self._id][ATTR_TEMP_COMFORT_C]):
```
Is this equivalent to something like this?
```
low = max(int(kwargs.get(ATTR_TARGET_TEMP_LOW), int(self._nobo.zones[self._id][ATTR_TEMP_COMFORT_C])
```

oyvindwe replied:
I must admin I didn't write this method originally, I only ported it to be async. I think the point is that it's not legal to set the low temperature higher than the high temperatur, and the other away around. However, both checks are against the _existing_ high/low temperatures.

Could the entire adjustment be skipped? I.e., is an HVAC component with support flag `SUPPORT_TARGET_TEMPERATURE_RANGE` guaranteed to always receive both `ATTR_TARGET_TEMP_LOW` and `ATTR_TARGET_TEMP_HIGH` , and that `low <= high`?

allenporter replied:
I am not sure: maybe manually test out with a service call and see what happens..

oyvindwe replied:
Made the method more robust (and readable).

allenporter commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,326 @@
+    # Find OFF command (week profile) to use for all zones:
+    command_off_name = config_entry.options.get(CONF_COMMAND_OFF)
+    command_on_by_id: dict[str, str] = {}  # By default, nothing can be turned on
+    if command_off_name in [None or ""]:
```
```suggestion
    if not command_off_name:
```

allenporter commented on homeassistant/components/nobo_hub/strings.json:
```diff
@@ -0,0 +1,66 @@
+        "data": {
+          "override_type": "Override type",
+          "command_off": "Week profile with 'off' setting",
+          "command_on_zone_1": "Normal week profile for zone 1",
```
High level question: Could it work to force the same profile name be used in every zone? Showing 20 different zone names seems not ideal

oyvindwe replied:
Good question, as this way to configure the integration really bothered me. Before implementing 32 zones here I asked the Norwegian community about their various setups. The answer is that is depends on the user's configuration. A week profile describes for when the heaters in a zone should be off, comfort, and eco mode, controlled by the hub. Some (e.g. I) have made different week profiles per zone, while others let HA control this, and use the same week profile for `normal` for all zones. Also, the limit of how many zones the hub supports is undocumented, but if you want to control each heater individually, 32 sounds like a reasonable limitation. (The limitation is really only in the config flow - this would be much prettier if it was possible to provide a dynamic length list in the config flow.)

Allowing different "normal" program per zone is also what's currently supported in the HACS version (using `configuration.yaml`), so removing that would make it harder for some users to convert. For more details, see https://github.com/echoromeo/hanobo

oyvindwe replied:
For some context - my current setup:
![image](https://user-images.githubusercontent.com/5984518/148702633-00eb3100-e2ba-4caa-a0db-1233381be532.png)

MartinHjelmare replied:
It's hard to understand why we do it like this. Why aren't we using a select entity per zone to select the profile for that zone?

Is there some more limitation that we need to consider when deciding the design for this than:
1. Zones
2. Profiles

What is the matrix of options that we need to represent with our user interface?

oyvindwe replied:
Configuring "normal" week profile per entity instead would be a nice solution, and clean up this mess.

Configuring the "off" week profile would be redundant to do on each entity, but it probably makes sense to do it the same place as the "normal" week profile is configured.

However, how do I build an option flow for a climate entity?

oyvindwe replied:
Here's my personal list of zones and week profiles:

Zones:
- Living room
- Bedrooms
- Guest room
- Study
- Basement living room
- Bathroom

Week profiles:
- Default
- Study
- Basement living room
- Guest room
- Bath
- Bedrooms
- Living room
- Home
- Away
- Basement living room home
- Bedrooms home
- Off

oyvindwe replied:
(Note the naming the normal week profile the same as the zone is just my personal naming convention.)

gjohansson-ST replied:
As per Martins comment would probably look nicer if there is a select entity per zone where you select the profile instead of running it through a config/options flow

oyvindwe replied:
@gjohansson-ST I agree, and I’m happy to implement that, but I need some pointers to where that should be plugged in and preferably some example where something similar is done. It doesn’t look like it’s possible to extend the entity configuration. The zones are represented as ClimateEntity instances.

MartinHjelmare replied:
We should not support setting the HVAC mode in the climate entity. The profile feature of the zone doesn't match what we represent in Home Assistant climate entities with the HVAC mode.

The profile setting feature and select platform should not be included in this PR. That should be a follow up PR.

Look at existing integrations that implement a `select.py` platform. Here are the docs:

https://developers.home-assistant.io/docs/core/entity/select

oyvindwe replied:
@MartinHjelmare The OptionFlow is not intended as a way to let the user switch between week profiles. It is a one-time setup to support HVAC mode to "off". This is based on the implementation in the HACS integration, where this is configured in `configuration.yaml`: https://github.com/echoromeo/hanobo/#there-is-no-override-in-nob%C3%B8-for-completely-off-but-there-is-a-workaround

MartinHjelmare replied:
We shouldn't abuse setting the HVAC mode when the device doesn't support the same concept. It doesn't matter what the custom integration does. Custom integrations often don't implement things correctly according to our architecture design.

oyvindwe replied:
Point taken. I'll remove this part of the configuration from the OptionFlow.

oyvindwe committed:
Update homeassistant/components/nobo_hub/config_flow.py

Review

Co-authored-by: Allen Porter <allen.porter@gmail.com>

oyvindwe commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,219 @@
+
+    def _prefill_identifier(self):
+        for (_, serial) in self.discovered_hubs:
+            return serial + "XXX"
```
The broadcast message from the hub contains the first 9 (of 12) digits of the hub's serial number. To connect, you have to provide the missing 3 digits (printed on the device), hence `XXX` needs to be replaced. This is explained in the `pynobo` documentation - I'm happy to include a comment here as well. The description for the field in `strings.json` may be more precise as well.

oyvindwe committed:
Update homeassistant/components/nobo_hub/climate.py

Review

Co-authored-by: Allen Porter <allen.porter@gmail.com>

oyvindwe committed:
Simplify if tests.

frenck commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,221 @@
+            step_id="user",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_SERIAL, default=default_suggestion): str,
+                    vol.Optional(CONF_IP_ADDRESS): str,
```
Is it serial or IP-based communication? I'm confused...

oyvindwe replied:
It's TCP/IP. To connect, you have to provide the device's full serial number (12 digits). The IP address and the first 9 digits of the serial number may be discovered on the local network. Should I rename the configuration key?

frenck replied:
Ah, check 👍 

So why do we ask for it? As in, we can always ask for it in a second step?

oyvindwe replied:
Serial number is always needed. IP address is optional, depending on your setup.

frenck replied:
> IP address is optional, depending on your setup.

Yes, that is what I meant. Can't we ask for IP if needed? (instead of always ask for it?)

oyvindwe replied:
I guess the case were you have multiple hubs with the same serial number prefix on your local network can be solved by trying to connect to them in order and see which accepts. But what if we discover a hub on the local network, but you want to connect to a remote hub that's not auto discovered (e.g. over a VPN connection)? I don't think it's a problem to offer an optional field together with a required field:

![image](https://user-images.githubusercontent.com/5984518/148824003-89cc4e70-0d61-4124-9333-cb8f173fb6b7.png)

Otherwise there are two cases for requesting IP address:
1. No hub is found on the local network
2. HA is not able to connect to any hub on the local network

Should these two cases have different flows then? For the first HA should require an IP address right away, while for the latter HA should report a problem and request serial and IP address? This adds a bit of complexity to the config flow.

allenporter replied:
I think making the common cases simpler from a user perspective is also worth the complexity when discovery finds devices to add.

oyvindwe replied:
Implemented @allenporter 's suggestion.

Config window when HA discovers a hub on local network:
<img width="604" alt="image" src="https://user-images.githubusercontent.com/5984518/179365370-f023953d-65b7-48b4-91fb-a58b5225a625.png">


Config window if no hub is discovered, or the user checked "Configure undiscovered hub":
<img width="605" alt="image" src="https://user-images.githubusercontent.com/5984518/179365390-8280389d-90c2-4edd-93b6-25b4e945625b.png">

oyvindwe committed:
Update homeassistant/components/nobo_hub/__init__.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

oyvindwe committed:
Update homeassistant/components/nobo_hub/__init__.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

oyvindwe committed:
Update homeassistant/components/nobo_hub/__init__.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

oyvindwe commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,54 @@
+    serial = entry.data.get(CONF_SERIAL)
+    ip_address = entry.data.get(CONF_IP_ADDRESS)
+    discover = ip_address is None
+    hub = nobo(serial=serial, ip=ip_address, discover=discover, loop=hass.loop)
```
This is to pass the event loop to the `pynobo` library, otherwise it calls `asyncio.get_event_loop()` and fails:
```
  File "/Users/omw/proj/home-assistant/homeassistant/components/nobo_hub/__init__.py", line 25, in async_setup_entry
    hub = nobo(serial=serial, ip=ip_address, discover=discover)
  File "/Users/omw/proj/home-assistant/venv/lib/python3.9/site-packages/pynobo.py", line 227, in __init__
    self._loop.run_until_complete(self.start())
  File "/Users/omw/.pyenv/versions/3.9.9/lib/python3.9/asyncio/base_events.py", line 618, in run_until_complete
    self._check_running()
  File "/Users/omw/.pyenv/versions/3.9.9/lib/python3.9/asyncio/base_events.py", line 578, in _check_running
    raise RuntimeError('This event loop is already running')
RuntimeError: This event loop is already running
```

frenck replied:
The library needs updating, as this will break Python 3.10

oyvindwe replied:
I see. I'll look into it, and follow up here as soon as a new version is released.

MartinHjelmare replied:
This still needs addressing.

oyvindwe committed:
Separate config step for manual configuration.

oyvindwe committed:
Separate config step for manual configuration.

oyvindwe committed:
Fixed indentation

oyvindwe committed:
Made async_set_temperature more robust

oyvindwe committed:
Thermometer supports tenths even though thermostat is in ones.

oyvindwe committed:
Preserve serial suffix in config dialog on error.

frenck commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,53 @@
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = hub
+
+    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
```
```suggestion
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
```

homeassistant commented:
Hi @GitHub-Action, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#50913).

Once you do that we will be able to review and accept this pull request.

Thanks!

oyvindwe commented:
(Apologies for the extra push that generated a lot of review requests. Seems that I managed to do a local pull on top of my rebase.)

oyvindwe committed:
Update homeassistant/components/nobo_hub/__init__.py

Co-authored-by: Franck Nijhof <frenck@frenck.nl>

oyvindwe committed:
Formatting (prettier)

oyvindwe committed:
Fix HA stop listener.

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,310 @@
+        )
+
+    @callback
+    def update(self) -> None:
```
It's not allowed to name the function `update` if it should run in the event loop. This is a polling entity that will use `update` to poll for new device data, running that on the executor thread pool. If we want it to run in the event loop we should use a coroutine function named `async_update`.

oyvindwe replied:
Thanks.

I think I messed up here when I realised that the device itself does not send an update when a zone changes mode based on a schedule stored on the device, hence HA needs to poll for this information. The call itself is handled locally by the library by checking the current time against the retrieved profiles. Any other state change is driven by a callback from the library.

I hope it's OK to do make this combination of polling and pushing. Otherwise the component needs a timer of its own that triggers every minute to push the state, and that seemed unnecessary to me when HA already supports polling periodically.

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,310 @@
+        self._override_type = override_type
+
+        # Register for callbacks before initial update to avoid race condition.
+        hub.register_callback(self._after_update)
```
Please register state update callbacks in `async_added_to_hass` entity method.

oyvindwe replied:
Thanks! Fixed, and I also added a deregister call to `async_will_remove_from_hass`. I assume that was the correct thing to do.

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,310 @@
+
+        # Register for callbacks before initial update to avoid race condition.
+        hub.register_callback(self._after_update)
+        self.update()
```
Pass the second argument `True` to `async_add_entities` instead.

oyvindwe replied:
Thanks! I wasn't aware of that possibility. Fixed and tested OK.

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,310 @@
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_devices: AddEntitiesCallback,
```
```suggestion
    async_add_entities: AddEntitiesCallback,
```

oyvindwe replied:
Thanks! Included in my own commit.

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,310 @@
+                _set_on_commands(command_on_by_id, command_on_dict, hub)
+
+    # Add zones as entities
+    async_add_devices(
```
```suggestion
    async_add_entities(
```

oyvindwe replied:
Thanks! Included in my own commit.

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,254 @@
+        def _hub(hub):
+            return hub[0], f"{hub[1]['serial_prefix']}XXX ({hub[1]['ip']})"
+
+        return OrderedDict(map(_hub, enumerate(self._discovered_hubs)))
```
Please use a dict comprehension instead.

oyvindwe replied:
Fixed.

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,254 @@
+            name = hub.zones[zone]["name"].replace("\xa0", " ")
+            on_command = (
+                on_commands[name]
+                if name in on_commands and on_commands[name] in profile_names
```
Please don't use ternary operator expression if it spans more than one line.

oyvindwe replied:
Fixed.

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,254 @@
+            schema = schema.extend(
+                {
+                    vol.Required(
+                        CONF_COMMAND_ON + "_zone_" + zone, default=on_command
```
Please use f-strings instead of string concatenation.

oyvindwe replied:
Fixed.

oyvindwe committed:
Review

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,252 @@
+    ) -> FlowResult:
+        """Handle manual configuration. Triggered if no devices are discovered or by user."""
+        errors = {}
+        serial, ip_address = None, None
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,252 @@
+        errors = {}
+        serial, ip_address = None, None
+        if user_input is not None:
+            serial = user_input.get(CONF_SERIAL)
+            ip_address = user_input.get(CONF_IP_ADDRESS)
```
```suggestion
            serial = user_input[CONF_SERIAL]
            ip_address = user_input[CONF_IP_ADDRESS]
```

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,252 @@
+            except NoboHubConnectError as error:
+                errors["base"] = error.msg
+
+        return self.async_show_form(
+            step_id="manual",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_SERIAL, default=serial): str,
+                    vol.Required(CONF_IP_ADDRESS, default=ip_address): str,
+                }
+            ),
+            errors=errors,
+        )
```
```suggestion
        user_input = user_input or {}
        return self.async_show_form(
            step_id="manual",
            data_schema=vol.Schema(
                {
                    vol.Required(CONF_SERIAL, default=user_input.get(CONF_SERIAL)): str,
                    vol.Required(
                        CONF_IP_ADDRESS, default=user_input.get(CONF_IP_ADDRESS)
                    ): str,
                }
            ),
            errors=errors,
        )
```

oyvindwe replied:
Why use `user_input.get(CONF_SERIAL)` here when you suggested to use `user_input[CONF_SERIAL]` above?

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,252 @@
+        )
+
+    async def _test_connection(self, serial, ip_address):
+        if serial is None or not len(serial) == 12 or not serial.isdigit():
```
`serial` can't be `None`.

oyvindwe replied:
Thanks! Fixed.

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,252 @@
+    async def async_step_init(self, user_input=None) -> FlowResult:
+        """Manage the options."""
+
+        hub = self.hass.data[DOMAIN][self.config_entry.entry_id].hub
```
```suggestion
        hub: nobo = self.hass.data[DOMAIN][self.config_entry.entry_id]
```

MartinHjelmare commented on tests/components/nobo_hub/test_config_flow.py:
```diff
@@ -0,0 +1,362 @@
+        unique_id="123456789012",
+        data={"serial": "123456789012"},
+    )
+    hass.data[DOMAIN] = {
```
Don't interact with integration details, like hass.data, in tests. Set up the integration while patching the library to avoid I/O. Then start the options flow.

https://developers.home-assistant.io/docs/development_testing#writing-tests-for-integrations

Note that we may remove the options flow for select entities.

oyvindwe replied:
Thanks! Fixed.

MartinHjelmare commented:
Please commit my suggestions directly in the github web UI so I don't need to review those changes again. You can batch multiple suggestions in a single commit by going to the `files` tab on this page.

oyvindwe commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,252 @@
+            return await self.async_step_manual()
+
+        errors = {}
+        serial_suffix = ""
```
If you want to configure a device that wasn't discovered on the local network (using UDP broadcast), we cannot require this field.

oyvindwe replied:
Not setting `serial_suffix` to empty string crashes when displaying when `default` is set. See below for comment.

MartinHjelmare replied:
We should then split this to a following step instead. The first step allows the user to select between configuring a discovered device and a manual device from the list of discovered devices including a "manual" device. Then the next step is either a step configuring one of the discovered devices or the step for a manual device.

oyvindwe replied:
@MartinHjelmare Got it! However, if I add a "manual" (or "undiscovered") device to the list, how do I localise the string value? The discovered hubs are displayed as `123456789XX: 1.2.3.4` (`123456789` is the serial prefix, `1.2.3.4` is the IP address).

MartinHjelmare replied:
Just call it "Manual".

oyvindwe replied:
OK. Fixed.

oyvindwe commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,252 @@
+
+        data_schema = vol.Schema(
+            {
+                vol.Optional("manual"): bool,
```
OK. Any suggestion to what this step should be named?

oyvindwe replied:
Forget above question - I misunderstood the comment (I read it as "manual" was a reserved step name).

oyvindwe replied:
Fixed.

oyvindwe commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,252 @@
+            {
+                vol.Optional("manual"): bool,
+                vol.Required("device"): vol.In(self._hubs()),
+                vol.Optional("serial_suffix", default=serial_suffix): str,
```
See above comment on why I didn't make this required. I added the default so in case of a typo you don't loose the value you entered.

oyvindwe replied:
Fixing this per above suggestion to add an extra step for discovered devices.

oyvindwe replied:
Fixed.

oyvindwe commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,252 @@
+            if "manual" in user_input:
+                return await self.async_step_manual()
+            serial, ip_address = None, None
+            if "serial_suffix" not in user_input or user_input["serial_suffix"] == "":
+                errors["base"] = "missing_serial_suffix"
+            else:
+                hub = self._discovered_hubs[user_input["device"]]
+                serial_prefix = hub["serial_prefix"]
+                serial_suffix = user_input["serial_suffix"]
+                serial = f"{serial_prefix}{serial_suffix}"
+                if "store_ip" in user_input and user_input["store_ip"]:
+                    ip_address = hub["ip"]
```
See below on discussion for keeping `serial_suffix` as optional.

oyvindwe replied:
Fixed in new config flow.

oyvindwe commented on homeassistant/components/nobo_hub/strings.json:
```diff
@@ -0,0 +1,75 @@
+          "device": "Discovered hubs",
+          "manual": "Configure undiscovered hub",
+          "serial_suffix": "Serial number suffix (3 digits)",
+          "store_ip": "Remember the IP address"
```
Hmm - that will add some complexity to the initialisation of the integration. I'll look into it.

oyvindwe replied:
It's not straight forward nor really necessary to implement periodic discovery. The discovery protocol only replies the serial prefix. To verify the serial suffix, it's necessary to connect to the hub, but the hub only supports 2 concurrent TCP/IP connections, one of these will be in use by the integration, and the other may be in use from e.g. a mobile app. If the IP address is the same, we can  be sure that it is the same hub, but in theory (however unlikely) there may be more than one hub on the local network, and they may have the the same serial number prefix. This corner case was the what "Remember the IP address" was intended to fix.

I agree the choice is a bit confusing, and the above corner case can also be supported by selecting the manual step, so I removed the check box for `store_ip`. The only drawback is that the user then must enter the entire serial number and the IP address, but I guess this rarely will be necessary.

The library itself does reconnect automatically, and if the hub has gotten a new IP address (e.g. from the local DHCP server), it will try to re-discover if the hub was auto-discovered in the first place.

I added always storing the IP address, and in addition a flag to say of the hub was discovered or not.

oyvindwe committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

oyvindwe commented:
> Please commit my suggestions directly in the github web UI so I don't need to review those changes again. You can batch multiple suggestions in a single commit by going to the `Files changed` tab on this page.


@MartinHjelmare Thank you very much for the review! I'll incorporate the rest of the comments this afternoon.

oyvindwe commented:
> @MartinHjelmare Thank you very much for the review! I'll incorporate the rest of the comments this afternoon.

I have most of the code in place, but I need a few afternoons to get the tests back in place before I can push.

oyvindwe committed:
Review
- Removed workaround to support "OFF" setting.
- Simplified config flow to add a new device.

oyvindwe commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,252 @@
+            errors=errors,
+        )
+
+    async def _create_configuration(self, serial, ip_address) -> FlowResult:
```
Fixed.

oyvindwe commented on homeassistant/components/nobo_hub/models.py:
```diff
@@ -0,0 +1,16 @@
+"""The nobo_hub integration models."""
```
Removed.

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+            },
+        )
+
+    async def _test_connection(self, serial: str, ip_address: str):
```
```suggestion
    async def _test_connection(self, serial: str, ip_address: str) -> str:
```

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+            raise NoboHubConnectError("invalid_serial")
+        try:
+            socket.inet_aton(ip_address)
+        except OSError:
```
```suggestion
        except OSError as err:
```

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+        try:
+            socket.inet_aton(ip_address)
+        except OSError:
+            raise NoboHubConnectError("invalid_ip") from OSError
```
```suggestion
            raise NoboHubConnectError("invalid_ip") from err
```

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+            }
+            return self.async_create_entry(title="", data=data)
+
+        override_type = self.config_entry.options.get(CONF_OVERRIDE_TYPE)
```
```suggestion
        override_type = self.config_entry.options.get(CONF_OVERRIDE_TYPE, CONF_OVERRIDE_TYPE_CONSTANT)
```

MartinHjelmare commented on homeassistant/components/nobo_hub/config_flow.py:
```diff
@@ -0,0 +1,212 @@
+            return self.async_create_entry(title="", data=data)
+
+        override_type = self.config_entry.options.get(CONF_OVERRIDE_TYPE)
+        if override_type != CONF_OVERRIDE_TYPE_NOW:
```
We can remove this.

oyvindwe replied:
Fixed.

MartinHjelmare commented on homeassistant/components/nobo_hub/const.py:
```diff
@@ -0,0 +1,15 @@
+CONF_AUTO_DISCOVERED = "auto_discovered"
+CONF_SERIAL = "serial"
+CONF_OVERRIDE_TYPE = "override_type"
+CONF_OVERRIDE_TYPE_CONSTANT = "Constant"
```
We don't prefix config values with `CONF_`, only config key constants.

MartinHjelmare replied:
```suggestion
OVERRIDE_TYPE_CONSTANT = "Constant"
```

MartinHjelmare commented on homeassistant/components/nobo_hub/const.py:
```diff
@@ -0,0 +1,15 @@
+CONF_SERIAL = "serial"
+CONF_OVERRIDE_TYPE = "override_type"
+CONF_OVERRIDE_TYPE_CONSTANT = "Constant"
+CONF_OVERRIDE_TYPE_NOW = "Now"
```
```suggestion
OVERRIDE_TYPE_NOW = "Now"
```

MartinHjelmare commented on homeassistant/components/nobo_hub/strings.json:
```diff
@@ -0,0 +1,44 @@
+        }
+      },
+      "manual": {
+        "description": "Configre a Nobø Ecohub not discovered on your local network. If your hub is on another network, you can still connect to it by entering the complete serial number (12 digits) and its IP address.",
```
```suggestion
        "description": "Configure a Nobø Ecohub not discovered on your local network. If your hub is on another network, you can still connect to it by entering the complete serial number (12 digits) and its IP address.",
```

MartinHjelmare commented on tests/components/nobo_hub/test_config_flow.py:
```diff
@@ -0,0 +1,275 @@
+
+async def test_configure_with_discover(hass: HomeAssistant) -> None:
+    """Test configure with discover."""
+    await setup.async_setup_component(hass, "persistent_notification", {})
```
We don't need to set up the persistent_notification integration anymore.

oyvindwe replied:
Fixed (in all tests).

MartinHjelmare commented on tests/components/nobo_hub/test_config_flow.py:
```diff
@@ -0,0 +1,275 @@
+    )
+    entry.add_to_hass(hass)
+
+    result = await hass.config_entries.options.async_init(entry.entry_id)
```
We need to set up the config entry before starting the option ns flow. Use `hass.config_entries.async_setup`. Make sure to patch the integration setup function.

oyvindwe replied:
Done.

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,208 @@
+    async def async_set_hvac_mode(self, hvac_mode: HVACMode) -> None:
+        """Set new target HVAC mode, if it's supported."""
+        if hvac_mode not in self.hvac_modes:
+            _LOGGER.warning(
```
Nowadays we prefer raising ValueError with the message as exception argument for bad user input. The message will be shown in the frontend.

oyvindwe replied:
Fixed.

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,208 @@
+        if hvac_mode == HVACMode.AUTO:
+            await self.async_set_preset_mode(PRESET_NONE)
+        elif hvac_mode == HVACMode.HEAT:
+            await self.async_set_preset_mode(PRESET_COMFORT)
```
What happens if the user changes preset mode to comfort directly? Does the device update the entity to heat mode?

oyvindwe replied:
Yes. Setting any preset mode directly will create an override make the zone go into `HVACMode.HEAT`.

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,208 @@
+        self._id = zone_id
+        self._nobo = hub
+        self._attr_unique_id = hub.hub_serial + ":" + zone_id
+        self._attr_name = hub.zones[self._id][ATTR_NAME]
```
Please set `_attr_has_entity_name` to True and set `_attr_device_info`.

https://developers.home-assistant.io/docs/core/entity?_highlight=_attr_has#entity-naming

oyvindwe replied:
Fixed `_attr_has_entity_name`.

I have a question regarding `_attr_device_info`: a _zone_ is a _group_ of devices (heaters). It is *not* possible to control each heater individually. The  serial number of each individual heater can be retrieved from the hub. Would it be more correct to register the hub itself and each individual heater as separate devices?

MartinHjelmare replied:
I'd make each zone its own device and also the hub a device.

https://developers.home-assistant.io/docs/device_registry_index#what-is-a-device

oyvindwe replied:
Done. I used the zone's name as suggested area - I think that is pretty likely to match for most users.

I didn't set manufacturer or model on the zone devices, as they are not physical entities.

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,208 @@
+        """Initialize the climate device."""
+        self._id = zone_id
+        self._nobo = hub
+        self._attr_unique_id = hub.hub_serial + ":" + zone_id
```
Please use f-strings instead of string concatenation.

oyvindwe replied:
Fixed.

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,208 @@
+    _attr_precision = PRECISION_TENTHS
+    _attr_preset_modes = PRESET_MODES
+    # Need to poll to get preset change when in HVACMode.AUTO.
+    _attr_should_poll = True
```
This is the default.
```suggestion
```

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,208 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+_ZONE_NORMAL_WEEK_LIST_SCHEMA = vol.Schema({cv.string: cv.string})
```
```suggestion
```

MartinHjelmare commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,61 @@
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Nobø Ecohub from a config entry."""
+
+    serial = entry.data.get(CONF_SERIAL)
```
```suggestion
    serial = entry.data[CONF_SERIAL]
```

MartinHjelmare commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,61 @@
+    """Set up Nobø Ecohub from a config entry."""
+
+    serial = entry.data.get(CONF_SERIAL)
+    discover = entry.data.get(CONF_AUTO_DISCOVERED)
```
```suggestion
    discover = entry.data[CONF_AUTO_DISCOVERED]
```

MartinHjelmare commented on homeassistant/components/nobo_hub/__init__.py:
```diff
@@ -0,0 +1,61 @@
+
+    serial = entry.data.get(CONF_SERIAL)
+    discover = entry.data.get(CONF_AUTO_DISCOVERED)
+    ip_address = None if discover else entry.data.get(CONF_IP_ADDRESS)
```
```suggestion
    ip_address = None if discover else entry.data[CONF_IP_ADDRESS]
```

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,208 @@
+        """Set new target temperature."""
+        low, high = None, None
+        if ATTR_TARGET_TEMP_LOW in kwargs:
+            low = int(kwargs.get(ATTR_TARGET_TEMP_LOW))  # type: ignore[arg-type]  # ATTR_TARGET_TEMP_LOW is float if set
```
```suggestion
            low = round(kwargs[ATTR_TARGET_TEMP_LOW])
```

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,208 @@
+        if ATTR_TARGET_TEMP_LOW in kwargs:
+            low = int(kwargs.get(ATTR_TARGET_TEMP_LOW))  # type: ignore[arg-type]  # ATTR_TARGET_TEMP_LOW is float if set
+        if ATTR_TARGET_TEMP_HIGH in kwargs:
+            high = int(kwargs.get(ATTR_TARGET_TEMP_HIGH))  # type: ignore[arg-type]  # ATTR_TARGET_TEMP_HIGH is float if set
```
```suggestion
            high = round(kwargs[ATTR_TARGET_TEMP_HIGH])
```

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,208 @@
+                low = min(low, high)
+            elif self.target_temperature_high is not None:
+                low = min(low, int(self.target_temperature_high))
+        elif high is not None and self.target_temperature_low is not None:
```
high can't be None of low is not None. So we should probably indent this case within the above block.

oyvindwe replied:
This is for the case if only `ATTR_TARGET_TEMP_HIGH` is passed in `kwargs`.

Can assume that either both or none of `ATTR_TARGET_TEMP_LOW` and `ATTR_TARGET_TEMP_HIGH` are set in `kwargs`?

Can we also assume that low <= high if they are set?

And even better, can we also assume that not having `TARGET_TEMPERATURE` is respected?

If so, we can simplify to:
```
        await self._nobo.async_update_zone(
            self._id,
            temp_comfort_c=round(kwargs[ATTR_TARGET_TEMP_HIGH]),
            temp_eco_c=round(kwargs[ATTR_TARGET_TEMP_LOW])
        )
```

I tried to look into the climate component code based on an earlier comment, but I didn't find any assertions of the above assumptions.

MartinHjelmare replied:
We can assume that both high and low target temps are passed in the service call if any of them is passed. The service schema validates this.

oyvindwe replied:
Thanks! Simplified using this assumption.

oyvindwe committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

oyvindwe committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

oyvindwe committed:
Fixed review comments

oyvindwe committed:
Update en.json with correction in review.

oyvindwe committed:
Implemented review comments:
- Register devices
- Simplifed async_set_temperature

MartinHjelmare commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -61,6 +75,17 @@ async def async_setup_entry(
 
+    # Register hub as device
+    dev_reg = device_registry.async_get(hass)
+    dev_reg.async_get_or_create(
```
Please do this from the init module entry setup.

oyvindwe replied:
Done.

oyvindwe committed:
Register hub as device in init module

MartinHjelmare commented on tests/components/nobo_hub/test_config_flow.py:
```diff
@@ -0,0 +1,271 @@
+        )
+        assert result["type"] == "form"
+        assert result["errors"] == {}
+        assert result["step_id"] == "manual"
```
Please continue each test flow to either a create entry, error or abort result.

oyvindwe committed:
Implemented review comments.
Upgraded pynobo to 1.4.0.

oyvindwe committed:
Apply suggestions from code review

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

oyvindwe committed:
Avoid tacking on the device name in the entity name

oyvindwe commented on homeassistant/components/nobo_hub/climate.py:
```diff
@@ -0,0 +1,209 @@
+        self._nobo = hub
+        self._attr_unique_id = f"{hub.hub_serial}:{zone_id}"
+        self._attr_name = hub.zones[self._id][ATTR_NAME]
+        self._attr_has_entity_name = False
```
I had to set `_attr_has_entity_name` to False to avoid entity names like "Kitchen Kitchen" (if the zone is named "Kitchen"). However, should we add e.g. "Heater" as a suffix or prefix to the entity names?

oyvindwe replied:
Changed after reading https://developers.home-assistant.io/blog/2022/07/10/entity_naming/  to 
```
self._attr_name = None
self._attr_has_entity_name = True
```
Seems to work as desired, but this is a bit confusing.

oyvindwe replied:
Or is this better?
```
self._attr_name = "Heater"
self._attr_has_entity_name = True
```
Then I get entity names like `Kitchen Heater`

Other alternatives are `Heater(s)` or `Heater zone`

MartinHjelmare replied:
If the entity represents the main feature of the device, the entity `_attr_name` should be set to `None` while `_attr_has_entity_name` is set to `True`.

oyvindwe committed:
Inherit entity name from device name

MartinHjelmare commented on homeassistant/components/nobo_hub/manifest.json:
```diff
@@ -3,7 +3,7 @@
   "config_flow": true,
   "documentation": "https://www.home-assistant.io/integrations/nobo_hub",
-  "requirements": ["pynobo==1.3.0"],
+  "requirements": ["pynobo==1.4.0"],
```
Side note: The event loop is not directly needed to create a task. Just use `asyncio.create_task`.

https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task

https://github.com/echoromeo/pynobo/blob/026e0c41776822a53b2c43546ebd4c2dd0d3b603/pynobo.py#L394-L395

oyvindwe replied:
Thanks - I will make a PR for `pynobo` with this simplification.

The implementation of `asyncio.create_task` is equivalent though, so it shouldn't be a blocker for this PR?

https://github.com/python/cpython/blob/main/Lib/asyncio/tasks.py#L371-L384

MartinHjelmare replied:
Ni, I just noticed it when looking at the library and wanted to mention it.

oyvindwe replied:
Is fixed and will be in the next release of `pynobo`.

oyvindwe commented:
Any ideo on why `test_browse_media_time[pyloop]` failed on Python 3.9 and 3.10? Should I just rebase on dev?

oyvindwe commented:
> Yeah, please try a rebase.

All checks passed after rebase. Anything else outstanding?

MartinHjelmare reviewed (approved):
Great!
